### DISCUSSION TITLE
What about  a DAX debugger for trainers and trainees ?

### DISCUSSION THREAD

**Discussion by didierterrien** [didierterrien/Power-BI-Sidetools#9] [Ideas]
Looking at this great video from Brian Grant, I was inspired to create DAX trainer, a dynamic and interactive way of teaching and practicing.
     [https://www.youtube.com/watch?v=b73kvjuJLv8](https://www.youtube.com/watch?v=b73kvjuJLv8)

It will be able to :
- save DAX formulas in history or in a queue. Formulas may content DAX debugger instructions in order to show one or more debug tables and to set them up
- play the queue step by step, go backward and forward
- display instructions to trainees for each step
The queue may be saved to a file and sent to trainees so they can play with it and adapt formulas seeing immediately results in DAX debugger.

More :
- The same training file can be used on a model selected by the trainer (training mode) and on a model selected by the trainee (quick measure mode)
- DAX trainer detects strings typed by trainees in formulas and provide information in popup (ex : do not use this function, better use this one)
- Screenshots record
- Trainee may send their modified formulas back to the trainer with comments
- Expected results may be displayed

Please submit more ideas and vote for DAX trainer.

**Comment by BrianGrantCSG** [comment]
I may just be a bit biased (just a bit) but I think this would be an absolutely fantastic tool for training and learning!

**Comment by didierterrien** [comment]
I created a first mockup. At this step, the example is very simple but the concept will accept much more complex formulas.
Please give feed back.

[DAX trainer mockup V1.0.pptx](https://github.com/didierterrien/Power-BI-Sidetools/files/6357571/DAX.trainer.mockup.V1.0.pptx)

**Reply by BrianGrantCSG** [reply]
I like the approach, I think one thing that might be helpful is including a couple of optional hints, perhaps 3 per measure. So if a trainee gets stuck or doesn't know where to start, they can click on "Hint 1" in the dialogue and the hint pops up in the user interface. Having a couple of hints would allow the trainee to start with a subtle hint, then optionally show a less subtle hint, then finally show a very obvious hint, or maybe even the finished version of the formula. 

I can see this being extremely helpful for people wanting to really practice common formulas (which I found very helpful when I got started).

**Reply by didierterrien** [reply]
Great idea ! I added a "?" button and modified the specification.
It might be very useful also for advanced trainees in order to read recommandations about performance, good practice, ...

[DAX trainer mockup V1.1.pptx](https://github.com/didierterrien/Power-BI-Sidetools/files/6397548/DAX.trainer.mockup.V1.1.pptx)

**Comment by gregoryjkramer88** [comment]
Does the debugger not install by default?

**Reply by didierterrien** [reply]
Hello Gregory,
Yes DAX debugger is installed with Power BI Sidetools since version 1.2.0

**Reply by gregoryjkramer88** [reply]
not sure what i'm doing wrong. why do i not see the icons on the far right of your tool with the paper and pencil icons. how do i get those?

**Reply by didierterrien** [reply]
It's because they are under test now. Many features are coming.
You can use DAX debugger without these icons.

**Reply by gregoryjkramer88** [reply]
ok, thanks...need to spend more time with the tool!

**Reply by gregoryjkramer88** [reply]
ok, in business now. watched the youtube video with you and Reid and see how to use the tool....very cool!

**Comment by didierterrien** [comment]
I added modes in the specification.
The same training file can be used :
    - on a model selected by the trainer (training mode)
    - on a model selected by the trainee (quick measure mode)

[DAX trainer mockup V1.2.pptx](https://github.com/didierterrien/Power-BI-Sidetools/files/7027472/DAX.trainer.mockup.V1.2.pptx)

**Comment by didierterrien** [comment]
More quick measures. Trainee can add in few clicks :
- DAX calendars
- Time intelligence calculation groups

[DAX trainer mockup V1.3.pptx](https://github.com/didierterrien/Power-BI-Sidetools/files/7082098/DAX.trainer.mockup.V1.3.pptx)
